contributors/design-proposals/federated-deployment.md

```diff
@@ -0,0 +1,127 @@
+
+## Introduction
+
+The purpose of this document is to provide a detailed design of how Kubernetes Deployments should be handled in Federation, with the special emphasis on rolling update support in deployment updates. 
+
+The internal design is based on [Federated ReplicaSet](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/federated-replicasets.md)
+
+## Background
+
+With Federation we are trying to provide Kubernetes objects and functionality spanning over multiple clusters. One of the key objects in Kubernetes is Deployment. Deployment allows users to maintain a defined number of pod replicas and conveniently update them if needed. The update functionality is the main thing that differentiate them from replica sets. ReplicaSets update everything at once while Deployments allow a slow, rolling update and rollback in case the update is unsuccessful. 
+
+## Requirements
+
+Federated deployments must provide the similar functionality as regular deployments, in particular :
+  + [R1] Work well with kubectl rollout
+    + History - provide all of the revisions of federated deployment.
+    + Pause/Resume - provide an ability to pause and resume an ongoing update
+    + Status - provide rollout status information
+    + Undo - provide go back to the previous version
+    + Require no/almost-no fixes in kubectl
+  + [R2] Allow rolling update
+    + One cluster at a time (preferred)
+    + All clusters at a time (additional)
+  + [R3] Be similar to what is currently implemented in regular Kubernetes.
+  + [R4] Allow similar scheduling extra features as Federated ReplicaSet
+    + Specifying min/max/weight for each of the clusters
+    + Be able to rebalance replicas once a cluster is added, lost or if there is no capacity.
+  + [R5] Each of the underlying clusters should be able to work independently if the federation control plane is down,
+  or there are some cross-cluster network issues. The clusters, when federation controll plane is down, shoule be as easily updatable/managable as if federation were present. For high availability, federation controll plane should be considered an "addon" that makes some extra features possible, not a component that brings or replaces the core K8S functionality. 
+  This basically means that Federated Deployment should create fully-functional deployments in the clusters. 
+
+## Design
+
+Regular Deployments don’t maintain the replicas by themselves. They let ReplicaSets keep track of them. Each of the deployments create a separate ReplicaSet for each of the pod template versions they have. To do a rolling upgrade the deployment slowly adds replicas to the replica set with the new pod template and takes them from the old one. The names of ReplicaSets are in form of deployment_name + “-” + podspec_hash. So it is relatively easy to tell which replicaset is responsible for pods with the given spec (if you are a machine of course). 
```

Comment:
As commented elsewhere, podtemplate hash is not stable across Kubernetes versions. We will need to change the naming scheme for ReplicaSets. The simplest solution is deployment.name + deployment.templategeneration. Or hash that and append to deployment.name but the new hash should differ in size from the old one so we can avoid conflicts with old replica sets that we cannot migrate.